README
Introduction
============

jQuery Plugin inspired by Lancelot for KDE no-click interface.

Automatically go to the link only with hover, no need to click!


Example
=======

	$('.lancelot').lancelot({
            hoverTime: 2000,		//time to launch the link in miliseconds
            aclass: "lancelotGo",	//style, @see lancelot.css
            atext: "",				//text to show
            show: "false",			//if true dont hide the lancelots
            speed: "fast",			//speed animation
            linkAction: "location",	//if "open" or "_blank" will try to open a new window
            atitle: "go in 2s",		//link title
            alink: false,			//if 'string' will use as the url, if 'function' will call to build the url
            element: "a",			//element to hover, ex: "span", "div"
            launch: false			//function to call when hoverTime is over
	});

Demo online: http://codexico.com.br/projetos/lancelot/demo/
Real use: http://ttlocal.info

Code for demo on github: http://github.com/codexico/jquery-plugin-lancelot-autolink


Advanced
--------
You can pass functions to "alink" and "launch", example:

    var googleSearch = function(obj){
        //obj is a reference to each $('.lancelot')
        var search = obj.text().replace(/^#/,'');
        return "http://www.google.com/search?q="+search;
    };

	var launcAction = function(obj){
		//do something
    };

	$('.lancelot').lancelot({
		alink: googleSearch,
        launch: launchAction
	});

This will create a link to google search with the text from $('.lancelot') and
execute launchAction when hoverTime is over.

See this working on http://ttlocal.info


License (choose one or more)
===========================

Creative Commons GNU General Public License License:
http://creativecommons.org/licenses/GPL/2.0/

MIT:
http://www.opensource.org/licenses/mit-license.php

GPL:
http://www.gnu.org/licenses/gpl.html


Contact
=======
twitter: @codexico
blog: http://codexico.com.br/blog/jquery-lancelot-plugin


Reference
=========

github:
http://github.com/codexico/jquery-plugin-lancelot-autolink
Project Backlog:
http://www.pivotaltracker.com/projects/104617

Lancelot links:
http://lancelot.fomentgroup.org
http://userbase.kde.org/Lancelot

Blog (portuguese):
http://codexico.com.br/blog/jquery-lancelot-plugin


"Copyleft; All Wrongs Reserved"
http://www.gnu.org/copyleft/copyleft.html

Changelog
Changelog
=========

 * Version 0.8
 * 2010-09-03
- function stop was unnecessary
- minified using yuicompressor
- compiled using Closure Compiler

 * Version 0.7.3
 * 2010-08-16
- JSLint

 * Version 0.7.2
 * 2010-08-16
- new parameter to 'launch': tryPop, will try to popup and if fails, redirect

 * Version 0.7.1
 * 2010-08-13
- Changelog
- jquery.lancelot.js documentation


 * Version 0.7
 * 2010-08-12
- using obj as reference to the external functions


 * Version 0.6
 * 2010-08-12
- new parameter "launch" //function to call when hoverTime is over


 * Version 0.5
 * 2010-08-12
- some code quality with JSLint
- playing with the Module Pattern


 * Version 0.4.1
 * 2010-08-12
- remove parameter "display"


 * Version 0.4
 * 2010-08-10
- new parameter "element", option to create other element instead of the default <a>


 * Version 0.3.1
 * 2010-08-06
- rename defaults


 * Version 0.3
 * 2010-08-04
- using images and css


 * Version 0.2.1
 * 2010-08-04
- better demo


 * Version 0.2
 * 2010-08-03
- demo
- timer
- README
- default parameters
- mousehover


 * Version 0.0.1
 * 2010-08-03
- commit to github

jquery.lancelot-compiled.js
(function(c){c.fn.lancelot=function(j){var a={};a.defaults={hoverTime:2E3,aclass:"lancelotGo",atext:"",show:"false",speed:"fast",linkAction:"location",atitle:"go in 2s",alink:false,element:"a",launch:false};a.o=c.extend(a.defaults,j);return this.each(function(){var d=c(this),b,e,g,h,i,f;b=d.attr("href");if(a.o.alink!==false){b=a.o.alink;if(c.isFunction(a.o.alink))b=a.o.alink(d)}a.ahref=b;a.o.hoverElement=' <a href="'+b+'" class="'+a.o.aclass+'" title="'+a.o.atitle+'">'+a.o.atext+"</a>";if(a.o.element!== "a")a.o.hoverElement=c(document.createElement(a.o.element)).addClass(a.o.aclass);d.append(a.o.hoverElement);e=d.find("."+a.o.aclass);a.goLink=e;if(a.o.show!=="true"){e.hide();d.hover(function(){e.fadeIn(a.o.speed)},function(){e.fadeOut(a.o.speed)})}f=function(){window.location.href=b};g=function(){(h=window.open(b,"ttlocal search"))?h.focus():f()};i=function(){switch(a.o.linkAction){case "open":window.open(b,"ttlocal search");break;case "_blank":window.open(b,"_blank");break;case "tryPop":g();break; default:f()}};a.t=false;e.hover(function(){a.t&&window.clearTimeout(a.t);a.t=a.o.launch!==false&&c.isFunction(a.o.launch)?window.setTimeout(a.o.launch(d),a.o.hoverTime):window.setTimeout(i,a.o.hoverTime)},function(){window.clearTimeout(a.t)})})}})(jQuery);

jquery.lancelot.js
/**
 * jQuery Lancelot Plugin
 * Allows navigation with mouse hover.
 *
 * Feel in use: http://ttlocal.info
 *
 * Author: codexico http://codexico.com.br
 *
 * Source: http://github.com/codexico/jquery-plugin-lancelot-autolink
 * Demo: http://codexico.com.br/projetos/lancelot/demo/
 * Project Backlog: http://www.pivotaltracker.com/projects/104617
 * blog (português): http://codexico.com.br/blog/jquery-lancelot-plugin
 *
 *
 * License (choose one or more):
 * Creative Commons GNU General Public License License: http://creativecommons.org/licenses/GPL/2.0/
 * MIT: http://www.opensource.org/licenses/mit-license.php
 * GPL: http://www.gnu.org/licenses/gpl.html
 * "Copyleft; All Wrongs Reserved": http://www.gnu.org/copyleft/copyleft.html
 *
 *
 * Version 0.8
 * 2010-09-03
 */

(function ($) {
    $.fn.lancelot = function (options) {

        //define a closure
        var L = {};


        L.defaults = {
            hoverTime: 2000,		//time to launch the link in miliseconds
            aclass: "lancelotGo",	//style, @see lancelot.css
            atext: "",			//text to show
            show: "false",		//if true dont hide the lancelots
            speed: "fast",		//animation. TODO: other options
            linkAction: "location",	//"location" only redirects, "open" or "_blank" will try to open a new window, "tryPop" will try to popup and then redirect.
            atitle: "go in 2s",		//link title
            alink: false,		//if 'string' will use as the url, if 'function' will call to build the url
            element: "a",		//element to hover, ex: "span", "div"
            launch: false               //function to call when hoverTime is over
        };
        L.o = $.extend(L.defaults, options);

        //the plugin
        return this.each(function () {
            //obj is a reference to each $('.lancelot')
            var obj = $(this),

            //the place where the Lancelot will take us
            ahref,

            //reference to the element created to be the Lancelot
            goLink,

            //function that try to open a new popup, and if cant, redirect the page
            tryPop,

            //reference to window opened
            c,

            //function to execute the Lancelot
            launch,

            //function that redirects the browser to another link
            redirectTo;


            //where we go?
            ahref = obj.attr("href");
            if (L.o.alink !== false) {
                ahref = L.o.alink;
                if ($.isFunction(L.o.alink)) {
                    ahref = L.o.alink(obj);
                }
            }
            L.ahref = ahref;


            //create and append element to hover
            L.o.hoverElement = ' <a href="' + ahref + '" class="' + L.o.aclass + '" title="' + L.o.atitle + '">' + L.o.atext + '</a>';
            if (L.o.element !== "a") {
                L.o.hoverElement = $(document.createElement(L.o.element)).addClass(L.o.aclass);
            }
            obj.append(L.o.hoverElement);


            //get the new created element
            goLink = obj.find("." + L.o.aclass);
            L.goLink = goLink;


            //show animation
            if (L.o.show !== "true") {
                goLink.hide();
                obj.hover(
                    function () {
                        goLink.fadeIn(L.o.speed);
                    },
                    function () {
                        goLink.fadeOut(L.o.speed);
                    }
                    );
            }

            redirectTo = function () {
                window.location.href = ahref;
            };


            //ALERT: works in some browsers... try before use in production
            tryPop = function () {
                c = window.open(ahref, "ttlocal search");
                if (c) {
                    c.focus();
                } else {
                    redirectTo();
                }
            };


            launch = function () {
                switch (L.o.linkAction) {
                    case "open":
                        window.open(ahref, "ttlocal search");
                        break;
                    case "_blank":
                        window.open(ahref, '_blank');
                        break;
                    case "tryPop":
                        tryPop();//TODO: accept parameters
                        break;
                    default:
                        redirectTo();
                }
            };


            //prepare to launch
            L.t = false;//timer
            goLink.hover(
                function () {
                    if (L.t) {
                        window.clearTimeout(L.t);
                    }
                    if (L.o.launch !== false && $.isFunction(L.o.launch)) {
                        L.t = window.setTimeout(L.o.launch(obj), L.o.hoverTime);//pass L now or it will be lost!
                    } else {
                        L.t = window.setTimeout(launch, L.o.hoverTime);
                    }
                },
                function () {
                    window.clearTimeout(L.t);
                }
                );


        });//each
    };//fn.lancelot
})(jQuery);